pkg/watchdog/watchdog.go

```diff
@@ -0,0 +1,228 @@
+//     wd, err := watchdog.Open(watchdog.Dev)
+//     while running {
+//         wd.KeepAlive()
+//     }
+//     wd.MagicClose()
+//
+// Open() arms the watchdog. MagicClose() disarms the watchdog.
+//
+// Note not every watchdog driver supports every function!
+//
+// For more, see:
+// https://www.kernel.org/doc/Documentation/watchdog/watchdog-api.txt
+package watchdog
+
+import (
+	"log"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// Dev is the name of the first watchdog. If there are multiple watchdogs, they
+// are named /dev/watchdog0, /dev/watchdog1, ...
+const Dev = "/dev/watchdog"
+
+// Various ioctl numbers.
+const (
+	wdiocGetSupport    = 0x80285700
+	wdiocGetStatus     = 0x80045701
+	wdiocGetBootStatus = 0x80045702
+	wdiocGetTemp       = 0x80045703
+	wdiocSetOptions    = 0x80045704
+	wdiocKeepAlive     = 0x80045705
+	wdiocSetTimeout    = 0xc0045706
+	wdiocGetTimeout    = 0x80045707
+	wdiocSetPreTimeout = 0xc0045708
+	wdiocGetPreTimeout = 0x80045709
+	wdiocGetTimeLeft   = 0x8004570a
+)
+
+// Status contains flags returned by Status() and BootStatus(). These are the
+// same flags used for Support()'s options field.
+type Status int32
+
+// Bitset of possible flags for the Status() type.
+const (
+	// Unknown flag error
+	StatusUnknown Status = -1
+	// Reset due to CPU overheat
+	StatusOverheat Status = 0x0001
+	// Fan failed
+	StatusFanFault Status = 0x0002
+	// External relay 1
+	StatusExtern1 Status = 0x0004
+	// ExStatusl relay 2
+	StatusExtern2 Status = 0x0008
+	// Power bad/power fault
+	StatusPowerUnder Status = 0x0010
+	// Card previously reset the CPU
+	StatusCardReset Status = 0x0020
+	// Power over voltage
+	StatusPowerOver Status = 0x0040
+	// Set timeout (in seconds)
+	StatusSetTimeout Status = 0x0080
+	// Supports magic close char
+	StatusMagicClose Status = 0x0100
+	// Pretimeout (in seconds), get/set
+	StatusPreTimeout Status = 0x0200
+	// Watchdog triggers a management or other external alarm not a reboot
+	StatusAlarmOnly Status = 0x0400
+	// Keep alive ping reply
+	StatusKeepAlivePing Status = 0x8000
+)
+
+// Option are options passed to SetOptions().
+type Option int32
+
+// Bitset of possible flags for the Option type.
+const (
+	// Unknown status error
+	OptionUnknown Option = -1
+	// Turn off the watchdog timer
+	OptionDisableCard Option = 0x0001
+	// Turn on the watchdog timer
+	OptionEnableCard Option = 0x0002
+	// Kernel panic on temperature trip
+	OptionTempPanic Option = 0x0004
+)
+
+type Watchdog struct {
+	f *os.File
+}
+
+func Open(dev string) (*Watchdog, error) {
+	f, err := os.OpenFile(dev, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &Watchdog{f: f}, nil
+}
+
+// Close closes the device without disarming the watchdog.
+func (w *Watchdog) Close() error {
+	return w.f.Close()
+}
+
+// MagicClose disarms the watchdog. However if the kernel is compiled with
+// CONFIG_WATCHDOG_NOWAYOUT=y, there may be no way to disarm the watchdog.
+func (w *Watchdog) MagicClose() error {
```

Comment:
Can Close just be == MagicClose? Why wouldn't one always do this?

Reply:
So:
* Close => Leaves the watchdog armed. Maybe you're doing something like a kexec and want the watchdog to stay armed while it happens. This is the default behavior for example if your kernel OOMs and kills the watchdogd process (which closes all the fds in the underlying driver).
* MagicClose => Disarms the watchdog. For example if you're going to kexec and need to disarm the watchdog.

I used kexec in both these examples. It depends on which policy you're after.

Reply:
I feel it would be nicer if the naming made that clear... Close and CloseAndDisarm? CloseArmed, CloseDisarmed?

Reply:
The problem if it is called CloseAndDisarm is it won't work as advertised if CONFIG_WATCHDOG_NOWAYOUT=y.
The term "MagicClose" isn't my quip. The feature is literally called "Magic Close" in the documentation: https://www.kernel.org/doc/Documentation/watchdog/watchdog-api.txt